Build a Temporal Application from scratch in TypeScript v2 (#345)
temporalio/temporal-learning: docs/getting_started/typescript/hello_world_in_typescript/index.md

  - Identify the four parts of a Temporal Workflow application.
  - Describe how the Temporal Server gets information to the Worker.
  - Observe how Temporal recovers from failed Activities.

:::

### Introduction

Creating reliable applications is a difficult chore. [Temporal](https://temporal.io/) lets you create fault-tolerant, resilient applications using programming languages you already know, so you can build complex applications that execute reliably and recover from failures.

In this tutorial, you will build your first Temporal Application from scratch using the [Temporal TypeScript SDK](https://github.com/temporalio/sdk-typescript). You'll develop a small application that asks for your name and then uses APIs to get your public IP address and your location based on that address. External requests can fail due to rate limiting, network interruptions, or other errors. Using Temporal for this application will let you automatically recover from these and other kinds of failures without having to write explicit error-handling code.

The app will consist of the following pieces:

> 1. Two [Activities](https://docs.temporal.io/activities): Activities are functions called by Workflows, and they contain any logic that might fail or behave differently at different times. The first Activity will get your IP address, and the second Activity will use that IP address to get your location.

Reviewer: I can see how "logic that behaves differently at different times" can perhaps be confusing to a first-time reader. How about something like, "and they contain logic that involves external systems or operations that may fail or produce different results depending on external conditions such as network availability."

1. A [Workflow](https://docs.temporal.io/workflows): Workflows are functions that define the overall flow of the application.  Your Workflow will execute both Activities you define. It will call the first Activity to fetch the IP address, and then use the result of that Activity to call the next Activity to find your location.

> 2. A [Worker](https://docs.temporal.io/workers): Workers host the Activity and Workflow code and executes the code piece by piece.

Reviewer 2: ```suggestion
2. A [Worker](https://docs.temporal.io/workers): Workers host the Activity and Workflow code and execute the code piece by piece.
```

> 3. A client program: Youll need to start your Worfklow. In this tutorial you'll create a small program using the Temporal Client to start the Workflow.

Reviewer: Do you want to link to a page on the Client since you linked Activity, Workflow, and Worker? https://docs.temporal.io/encyclopedia/temporal-sdks#temporal-client

Author: ```suggestion
3. A client program: Youll need to start your Worfklow. In this tutorial you'll create a small program using the [Temporal Client](https://docs.temporal.io/encyclopedia/temporal-sdks#temporal-client) to start the Workflow.
```

> 3. A client program: Youll need to start your Worfklow. In this tutorial you'll create a small program using the Temporal Client to start the Workflow.

Reviewer: ```suggestion
3. A client program: You'll need to start your Workflow. In this tutorial you'll create a small program using the Temporal Client to start the Workflow.
```

Author: Fixed in previous comment.

```json
"dependencies": {
  "@temporalio/activity": "^1.11.5",
  "@temporalio/client": "^1.11.5",
  "@temporalio/worker": "^1.11.5",
  "@temporalio/workflow": "^1.11.5",
  "nanoid": "3.x"
},
```

You'll use Node.js's build-in `fetch` library in your application. To use it with TypeScript, open the file `tsconfig.json` and locate the `"lib"` key under `"compilerOptions"`and add the `DOM` library to the array:

```json
"compilerOptions": {
 ...
  "lib": ["es2020","DOM"],
 ...
}
```

This ensures that `fetch` will be available as a global module.

With the project created, you'll create the application's core logic.

## Write functions to call external services

Your application will make two HTTP requests. The first request will return your current public IP, while the second request will use that IP to provide city, state, and country information.

> You'll use Temporal Activities to make these requests. You use Activities in your Temporal Applications to execute non-deterministic code or perform operations that may fail.

Reviewer: I think it may be beneficial to explain WHY you use Activities to make these external requests, because a first-time reader would perhaps wonder why we don't do this in Workflows. 

Consider adding something like this:

"One of the key benefits of using Activities is their built-in fault tolerance. If an Activity fails, Temporal can automatically retry it until it succeeds or reaches a specified retry limit. This ensures that transient issues, like network glitches or temporary service outages, don't result in data loss or incomplete processes."

Reviewer: This also gives a preview before line 212. Because if I read line 212 without an introduction to Retry Policies first, I would wonder what that meant.

Author: ```suggestion
You'll use Temporal Activities to make these requests. You use Activities in your Temporal Applications to execute non-deterministic code or perform operations that may fail.

If an Activity fails, Temporal can automatically retry it until it succeeds or reaches a specified retry limit. This ensures that transient issues, like network glitches or temporary service outages, don't result in data loss or incomplete processes.
```

Open `src/activities.ts` and remove all the existing code in the file. You'll replace it with your own.

Add the following code to define a Temporal Activity that retrieves your IP address from `icanhazip.com`:

```ts
// Get the IP address
export async function getIP(): Promise<string> {
  const response = await fetch('https://icanhazip.com');
  const data = await response.text();
  return data.trim();
}
```

This function looks like a regular TypeScript function. With the Temporal TypeScript SDK, you define Activities using an exportable TypeScript module.

The response from `icanhazip.com` is plain-text, and it includes a newline, so you trim off the newline character before returning the result.

Notice that there's no error-handling code in this function. When you build your Workflow, you''ll use Temporal's Activity Retry policies to retry this code automatically if there's an error.

> Now add the second Activity that accepts an IP address and retrives location data. In In `src/activities.ts`, add the following code to define it:

Reviewer: ```suggestion
Now add the second Activity that accepts an IP address and retrieves location data. In `src/activities.ts`, add the following code to define it:
```

```ts
// Use the IP address to get the location.
export async function getLocationInfo(ip: string): Promise<string> {
  const url = `http://ip-api.com/json/${ip}`;
  const response = await fetch(url);
  const data = await response.json();
  return `${data.city}, ${data.regionName}, ${data.country}`;
}
```

This Activity follows the same pattern as the `getIP` Activity. It's an exported async function that uses `fetch` to call a remote service. This time, the service returns JSON data rather than text.

> While Activities can accept input argument, it's a best practice to send a single argument rather than multiple arguments. In this case you only have a single String. If you had more than one argument, you should bundle them up in a serializable object.  Review the [Activity parameters](https://docs.temporal.io/dev-guide/typescript/foundations/#activity-parameters) section of the Temporal documentation for more details, as there are some limitations you'll want to be aware of when running more complex applications.

Reviewer 2: ```suggestion
While Activities can accept input arguments, it's a best practice to send a single argument rather than multiple arguments. In this case you only have a single String. If you had more than one argument, you should bundle them up in a serializable object.  Review the [Activity parameters](https://docs.temporal.io/dev-guide/typescript/foundations/#activity-parameters) section of the Temporal documentation for more details, as there are some limitations you'll want to be aware of when running more complex applications.
```

You've created your two Activities. Now you'll coordinate them using a Temporal Workflow.

## Control application logic with a Workflow

Workflows are where you configure and organize the execution of Activities. You define a Workflow by writing a *Workflow Definition* using one of the Temporal SDKs. Review the [Develop Workflows](https://docs.temporal.io/develop/typescript/core-application#develop-workflows) section of the Temporal documentation for more about Workflows in TypeScript.

> Open the file `src/workflows.src` and remove the code in the file since you'll add your own. Then add the following code to import the Activities and configure how the Workflow should handle failures:

Reviewer: I would add a link to the different retry policies since there isn't too much of an explanation of what they are for.

Consider something like this:

```suggestion
Open the file `src/workflows.src` and remove the code in the file since you'll add your own. Then add the following code to import the Activities and configure how the Workflow should handle failures with a [retry policy](https://docs.temporal.io/encyclopedia/retry-policies).
```

```typescript
import * as workflow from '@temporalio/workflow';

// Only import the activity types
import type * as activities from './activities';

// Load Activities and assign the Retry Policy
const { getIP, getLocationInfo} = workflow.proxyActivities<typeof activities>({
  retry: {
    initialInterval: '1 second',
    maximumInterval: '1 minute',
    backoffCoefficient: 2,
    // maximumAttempts: 5,
  },
  startToCloseTimeout: '1 minute',
});

> ```

Reviewer: I would leave a comment as to which each retry policy is.

```ts
// Only import the activity types
import type * as activities from './activities';
import { proxyActivities } from '@temporalio/workflow';

// Load Activities and assign the Retry Policy
const { getIP, getLocationInfo} = proxyActivities<typeof activities>({
  retry: {
    initialInterval: '1 second', //amount of time that must elapse before the first retry occurs
    maximumInterval: '1 minute', //maximum interval between retries
    backoffCoefficient: 2, //how much the retry interval increases
    // maximumAttempts: 5, //maximum number of execution attempts that can be made
  },
  startToCloseTimeout: '1 minute', //maximum time allowed for a single Activity Task Execution
});
```

> The Temporal TypeScript SDK requires that Workflows and Activities run in separate environments. Temporal Workflows must be deterministic, and so Workflows are run in a sandbox that checks code for determinism. Since you run your non-deterministic operations in Activities, you configure your Workflow to call Activities through a proxy. That's why you import their types rather than the functions themselves.

Reviewer: Maybe link to the page about deterministic constraints because we aren't explaining here WHY workflows must be deterministic

https://docs.temporal.io/workflows#deterministic-constraints

> The Temporal TypeScript SDK requires that Workflows and Activities run in separate environments. Temporal Workflows must be deterministic, and so Workflows are run in a sandbox that checks code for determinism. Since you run your non-deterministic operations in Activities, you configure your Workflow to call Activities through a proxy. That's why you import their types rather than the functions themselves.

Author: ```suggestion
The Temporal TypeScript SDK requires that Workflows and Activities run in separate environments. Temporal Workflows [must be deterministic](https://docs.temporal.io/workflows#deterministic-constraints) so that Temporal can replay your Workflow in the event of a crash,  and the TypeScript SDK runs Workflows in a sandbox that checks code for determinism to enforce this.

Since you run your non-deterministic operations in Activities, you configure your Workflow to call Activities through a proxy. That's why you import their types rather than the functions themselves.
```

Reviewer: LGTM

> The `proxyActivities` method is also where you set options for how Temporal works with Activities. In this example, you have specified that the Start-to-Close Timeout for your Activity will be one minute, meaning that your Activity has one minute to begin before it times out. Of all the Temporal timeout options, `startToCloseTimeOut` is the one you should always set.

Author: yes! great catch!

You also set the Retry Policy for Activities this way. In this example, you're using the default Retry Policy values, but by leaving it in you have a more clear picture of what happens. Note that the `maximumAttempts` is commented out, which means there's no limit to the number of times Temporal will retry your Activities if they fail.

With the imports and options in place, you can define the Workflow itself. In the TypeScript SDK, you implement a Workflow the same way you define an Activity; using an exportable async TypeScript function. Add the following code to call both Activities, using the value of the first as the input to the second:

```ts
// The Temporal Workflow.
export async function getAddressFromIP(name: string): Promise<string> {

  try {
    const ip = await getIP();
    try {
      const location = await getLocationInfo(ip);
      return `Hello, ${name}. Your IP is ${ip} and your location is ${location}`;
    } catch (e) {
      throw new workflow.ApplicationFailure("Failed to get location");
    }
  } catch (e) {
    throw new workflow.ApplicationFailure("Failed to get IP");
  }

}
```

> This code uses a `try/catch` block, but because you've configured unlimited retries, there won't be any exceptions caught. However, if you change the Retry Policy's maximum retries, or you specify specific non-retryable exceptions, this code will be in place to handle those errors.

Reviewer: ```suggestion
This code uses a `try/catch` block, but because you've configured unlimited retries, there won't be any exceptions caught. However, if you change the Retry Policy's maximum retries, or you specify non-retryable exceptions, this code will be in place to handle those errors.
```

Next you'll create a Worker that executes the Workflow and Activities.

## Configure and run a Worker

> When you start a Temporal Workflow, the Workflow and its Activities get scheduled on the Temporal Service's [Task Queue](https://docs.temporal.io/concepts/what-is-a-task-queue).  A [Worker](https://docs.temporal.io/concepts/what-is-a-worker) hosts Workflow and Activity functions and polls the Task Queue for tasks related to those Workflows and Activities. After the Worker runs the code, it communicates the results back to the Temporal Service where they're stored in the Event History.

Reviewer: I would add a sentence to this to explain WHY the results are stored in the Event History. Consider adding something like:

```suggestion
When you start a Temporal Workflow, the Workflow and its Activities get scheduled on the Temporal Service's [Task Queue](https://docs.temporal.io/concepts/what-is-a-task-queue). A [Worker](https://docs.temporal.io/concepts/what-is-a-worker) hosts Workflow and Activity functions and polls the Task Queue for tasks related to those Workflows and Activities. After the Worker runs the code, it communicates the results back to the Temporal Service where they're stored in the Event History. This records the Workflow's entire execution, enabling features like fault tolerance by allowing the Workflow to replay in case of Worker crashes or restarts.
```

To define a Worker, you use the Temporal SDK to define a Worker Program.

> You'll need to specify the name of the Task Queue in your Worker Program, and any time you interact with a Workflow from a client application. You specify the Task Queue name as a case-insensitive string. Since you'll use this string in multiple places, create a constant to guarantee you use it consistently.

Reviewer: I personally felt the first sentence a little unclear. How about something like:

```suggestion
In your Worker Program, you need to specify the name of the Task Queue, which must match the Task Queue name used whenever you interact with a Workflow from a client application. This ensures that the Worker processes tasks for the correct Workflow. The Task Queue name is a case-insensitive string. To ensure consistency and avoid errors, define the Task Queue name as a constant that can be reused throughout your code.
```

Author: love it.

    //
    // The worker runs until it encounters an unexpected error or the process receives a shutdown signal registered on
    // the SDK Runtime object.
    //
    // By default, worker logs are written via the Runtime logger to STDERR at INFO level.
    //
    // See https://typescript.temporal.io/api/classes/worker.Runtime#install to customize these defaults.
    await worker.run();
  } finally {
    // Close the connection once the worker has stopped
    await connection.close();
  }
}

run().catch((err) => {
  console.error(err);
  process.exit(1);
});

```

The code imports the `TASK_QUEUE_NAME` constant along with all of the Activities in the `src/activities.ts` file. It then defines an async function named `run` which creates and runs a Worker. The Worker takes a configuration object that specifies a  `workflowsPath` (the location of your Workflow file), your Activity functions, and the name of the Task Queue.

In this case your Worker will run your Workflow and your two Activities, but there are cases where you could configure a Worker to run Activities, and another Worker to run the Workflows.

> Now you will use an `npm` script to start your Worker using Nodemon, which will reload whenever it detects changes in your file, hence the command name `start.watch`. Be sure you have started the local Temporal Service and execute the following command to start your Worker:

Reviewer: Broke up line 357 a bit as it was a bit long:

```suggestion
Now you will use an npm script to start your Worker with Nodemon. Nodemon automatically reloads whenever it detects changes in your file, hence the command name `start.watch`. Be sure you have started the local Temporal Service and execute the following command to start your Worker:
```

```command
npm run start.watch
```

The Worker runs and you see the following output:

```output

> > temporal-hello-world@0.1.0 start.watch

Author: Great catch - I forgot to replace this during testing!

> nodemon src/worker.ts

[nodemon] 2.0.22
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): src/**/*
[nodemon] watching extensions: ts
[nodemon] starting `ts-node src/worker.ts`
2024-12-09T18:27:38.879Z [INFO] Creating worker {
  sdkComponent: 'worker',
  taskQueue: 'ip-address-ts',
  options: {
   ...
  }
}
...

2024-12-09T18:27:39.412Z [INFO] Workflow bundle created { sdkComponent: 'worker', taskQueue: 'ip-address-ts', size: '0.82MB' }
2024-12-09T18:27:39.493Z [INFO] Worker state changed {
  sdkComponent: 'worker',
  taskQueue: 'ip-address-ts',
  state: 'RUNNING'
}
```

In the output, you see the Worker options and their values, the Webpack bundling output, and the state of the Worker.

> Before running this Workflow, write tests to prove it works as expected.

Author: Yes that's correct. But I'll clarify.

## Write tests to ensure things work

The Temporal TypeScript SDK includes functions that help you test your Workflow executions. Let's add a basic unit test to the application to make sure the Workflow works as expected.

> You'll use the [mocha](https://mochajs.org/) package to build your test cases and mock the Activity so you can test the Workflow in isolation. You'll also use the `@temporalio/testing` package which will download a test server which provides a `TestWorkflowEnvironment`.

Reviewer 2: ```suggestion
You'll use the [mocha](https://mochajs.org/) package to build your test cases and mock the Activity so you can test the Workflow in isolation. You'll also use the `@temporalio/testing` package, which will download a test server that provides a `TestWorkflowEnvironment`.
```

Running the tests requires using the `mocha` command along with requiring the following libraries and pointing the test runner to the appropriate folder. Here's what the command would look like to run the tests:

```command
npx  mocha \
  --require ts-node/register \
  --require source-map-support/register \
  src/mocha/*.test.ts
```

However, since this is quite verbose to type into the command line every time, you'll find a `test` script defined in `package.json` that runs this command for you.

Ensure you've saved all of your files and execute your tests with the following command:

```command
npm test
```

The test environment starts, spins up a Worker, and executes the Workflow in the test environment. At the end, you'll see that your test passes:

```output
    ✓ successfully completes the Workflow with mocked Activities (543ms)

  1 passing (723ms)
```

> :::tip

Reviewer 2: _Nice_ tip; that's exactly what happened.

If you get an error that your test has timed out, run it again. The first time you run the tests on your local machine, the test server might not start quickly enough.
:::

With a Workflow test in place, you can write unit tests for the Activities.

> Both of your Activities make external calls, and it'll be difficult to write tests that reliably test your Activities if they make calls to external services because you'll get a different IP address based on where your machine is located. To ensure the Activities are testable in isolation, you'll stub out the `fetch` calls using the sinon library.

Reviewer: Broke this up a bit:

```suggestion
Both of your Activities make external calls. Testing these Activities reliably can be challenging if they depend on external services. For example, the IP address may vary based on your machine's location.

To make the Activities testable in isolation, you’ll stub out the fetch calls using the `sinon` library.
```

> Both of your Activities make external calls, and it'll be difficult to write tests that reliably test your Activities if they make calls to external services because you'll get a different IP address based on where your machine is located. To ensure the Activities are testable in isolation, you'll stub out the `fetch` calls using the sinon library.

Reviewer 2: Do you want to add any description of sinon, even if it's just "sinon is a JavaScript testing library that provides mocks for situations like this one"?

> Add `sinon` as a development dependency, along with its type definitions:

Reviewer: Since this is its own standalone step, I would explain what `sinon` is for. Consider adding something like:

We will use the `sinon` library for creating mocks and stubs. Stubs can be seen as a dummy implementation of a method or object in a test. You can use a stub to force a method to throw an error in a function in order to test error handling. They do not directly simulate code functionality.

> Add `sinon` as a development dependency, along with its type definitions:

Author: ```suggestion
Sinon.js is a JavaScript testing library that provides powerful tools like spies, stubs, and mocks, allowing you to replace dependencies with controlled, testable behavior.

Add `sinon` as a development dependency, along with its type definitions:
```

Reviewer: LGTM

```command
npm i sinon @types/sinon --save-dev
```

Create the file `src/mocha/activities.test.ts` and add the following code to import the testing libraries you'll use:

```typescript
import { MockActivityEnvironment } from '@temporalio/testing';
import { describe, it } from 'mocha';
import * as activities from '../activities';
import assert from 'assert';
import sinon from 'sinon';
```

> The `MockActivityEnvironment `from the `@temporalio/testing` package lets you test Activities as if they were part of a Temporal Application.

Reviewer: ```suggestion
The `MockActivityEnvironment` from the `@temporalio/testing` package lets you test Activities as if they were part of a Temporal Application.
```

> The `MockActivityEnvironment `from the `@temporalio/testing` package lets you test Activities as if they were part of a Temporal Application.

Reviewer 2: ```suggestion
The `MockActivityEnvironment` from the `@temporalio/testing` package lets you test Activities as if they were part of a Temporal Application.
```

To test the `getLocation` Activity, you use a similar approach. Add the following code to the `src/mocha/activities.ts` file:

```typescript
describe('getLocation activity', async () => {
  it('successfully gets the location', async () => {
    const ip = '123.45.67.89';
    const fakeLocation = {
      city: 'Sample City',
      regionName: 'Sample Region',
      country: 'Sample Country'
    };

    const stub = sinon.stub(global, 'fetch').resolves({
      json: () => Promise.resolve(fakeLocation),
    } as Response);

    try {
      const env = new MockActivityEnvironment();
      const location = await env.run(activities.getLocationInfo, ip);
      assert.strictEqual(location, `${fakeLocation.city}, ${fakeLocation.regionName}, ${fakeLocation.country}`);
    } finally {
      stub.restore();
    }
  });
});
```

This test looks similar to the previous test; you stub out the `fetch` method and ensure it returns the expected data, and then you execute the Actiity in the `MockActivityEnvironment`. Then you restore the stubbed `fetch` method.

Run the tests again to see them pass:

```command
npm test
```

> Now that you have your tests passing, it's time for you to start a Workflow Execution.

Reviewer 2: ```suggestion
Now that you have your tests passing, it's time to start a Workflow Execution.
```

- You'll need to add code that lets you specify the person's name as the command-line argument so it can be passed to the Workflow, instead of using the hard-coded value of `Temporal`.

Open the `src/client.ts` file in your editor and modify it so it looks like the following:

```typescript
import { Connection, Client } from '@temporalio/client';
import { getAddressFromIP } from './workflows';
import { nanoid } from 'nanoid';
import { TASK_QUEUE_NAME } from './shared';
import process from 'process';

async function run() {

  if (process.argv.length <= 2) {
    console.error('Must specify a name as the command-line argument');
    process.exit(1);
  }

  const name = process.argv[2];
  const connection = await Connection.connect({ address: 'localhost:7233' });
  const client = new Client({ connection });

  const handle = await client.workflow.start(getAddressFromIP, {
    taskQueue: TASK_QUEUE_NAME,
    args: [name],
    workflowId: 'getAddressFromIP-' + nanoid(),
  });
  console.log(await handle.result());
}

run().catch((err) => {
  console.error(err);
  process.exit(1);
});

```

You import the `process` module so you can access command-line arguments you'll pass.

> In the `run` function you check to see if there is at least two arguments passed. You need to check for two arguments because `process.argv` includes the Node.js executable path and the script path as the first two elements, so the actual command-line arguments start at index 2.

Reviewer 2: ```suggestion
In the `run` function you check to see if there are at least two arguments passed. You need to check for two arguments because `process.argv` includes the Node.js executable path and the script path as the first two elements, so the actual command-line arguments start at index 2.
```

> The `run` function then sets up a connection to your Temporal Server, invokes your Workflow, passes in an argument for the `name`, and assigns the Workflow a unique identifier with Nanoid. The client dispatches the Workflow on the same `hello-world` Task Queue that the Worker is polling on.

Reviewer 2: I don't think the Task Queue will be 'hello-world'. As I mentioned above, I believe it's 'empty'.

:::note Specify a Workflow Id

A Workflow ID is unique in a Namespace and identifies a Workflow Execution. Using an identifier that reflects some business process or entity is a good practice. For example, you might use a customer identifier as part of the Workflow Id if you run one Workflow per customer. This would make it easier to find all of the Workflow Executions related to that customer later.

In this tutorial you're generating a UUID and appending it to a string that identifies the Workflow.
:::

With the script created, open a new terminal and switch to the project directory:

```command

> cd ip-geo-ts

Reviewer 2: Should be `cd temporal-ip-geolocation`

```command
npm run workflow Brian
```

You'll see the following output:

```output
> temporal-hello-world@0.1.0 workflow
> ts-node src/client.ts Brian

Hello, Brian. Your IP is 198.51.100.23 and your location is Seattle,  Washington, United States
```

Your Temporal application works. Now review it in the Temporal Web UI.

## Exploring your application in the Web UI

The Temporal Web UI gives you insights into your Workflow's execution. Open the Temporal Web UI by visiting `http://localhost:8233` and click on your completed Workflow to view the execution history. You'll see results similar to the following image:

![The results of the Workflow](images/overview.png)

You'll see the dates the Workflow Exeuction ran, how long it took to run, the input to the Workflow, and the result.

Below that, you see the Event History, detailing the entire flow, including the inputs and outputs of the Activity Executions:

![The Workflow History](images/history.png)

The most recent event is at the top, so read the history from the bottom up to see each step in the process. Using this history, you can see exactly how your Workflow executed and pinpoint any places things might have gone wrong.

> Temporal stores the results of each Activity in this history, as you can see im the image. If there was a system crash between the `getIP` and `getLocationInfo` Activity Executions, a new Worker would re-run the Workflow, but would use the previous Event History to reconstruct the Workflow's state. Instead of re-running the `getIP` function, it would use the previous run's value and continue on.

Reviewer: I would re-emphasize why this is important. Consider adding something like at the end:

This is important in the case of failure because it prevents duplicate external calls or side effects, such as calling the same API multiple times or generating inconsistent results. By relying on the stored Event History, Temporal ensures that the Workflow can recover seamlessly, maintaining reliability and consistency even after a crash.

Author: Isn't this covered in the next paragraph?

Reviewer: I like emphasis lol. It introduces the concept here and then re-emphasizes as you get to the next part. Up to you though!

Next, you'll explore how Temporal handles failed Activities.

## Observe automatic retries

When you developed your Activities, you didn't include any error-handling code. So if there's a problem making the request, the Workflow will handle the error using the Retry Policy.

Test this out. Disconnect your local machine from the Internet by turning off your Wi-Fi connection or unplugging your network cable.

Then, with the local Temporal Service running and your Worker running, switch to the Terminal window where you ran your Workflow and run it again:

```command
npm run workflow
```

This time you don't get a response.

Visit `http://localhost:8233` to open the Temporal Web UI and locate the Workflow Execution that's currently running. When you select it, you'll see something like the following image, indicating that there's a problem:
![timeline](images/timeline_failed.png)

As you can see, the `getIP` Activity has failed and Temporal is retrying it. Scroll down to the Event History and you'll see the failure represented there:

![Failed Activity](images/history_failed.png)

> Select the **Pending Activity** in the table to see why it failed and you'll see the stack trace:

Reviewer 2: ```suggestion
Select the **Pending Activity** item in the table to see why it failed and you'll see the stack trace:
```

![activity_stack_trace](images/activity_stack_trace.png)

In the **Last Failure** field, you can see there was a TypeScript error indicating that `fetch` failed.

Connect to the Internet again and wait. After a few moments, the Workflow recovers and completes:

![history_recovered](images/history_recovered.png)

If you return to your terminal where you launched the Workflow, you'll find your results there as well.

You can recover from failures by letting Temporal handle them for you instead of writing complex error-handling logic. You can also decide that you only want to retry a fixed number of times, or that you only want to recover on certain kinds of errors.

## Conclusion

> In this tutorial you built your first Temporal Application, using TypeScript functions to build a resiliant application that recovered from failure. You wrote tests to verify that it works and reviewed the Event History for a working execution. You also tested your Workflow without an Internet connection to understand how Temporal recovers from failures like network outages.

Reviewer: ```suggestion
In this tutorial you built your first Temporal Application, using TypeScript functions to build a resilient application that recovered from failure. You wrote tests to verify that it works and reviewed the Event History for a working execution. You also tested your Workflow without an internet connection to understand how Temporal recovers from failures like network outages.
```

Now you will use an `npm` script to start your Worker with Nodemon. Nodemon automatically reloads whenever it detects changes in your file, hence the command name `start.watch`. Be sure you have started the local Temporal Service and execute the following command to start your Worker:

```command
npm run start.watch
```

The Worker runs and you see the following output:

```output
> temporal-hello-world@0.1.0 start.watch
> nodemon src/worker.ts

[nodemon] 2.0.22
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): src/**/*
[nodemon] watching extensions: ts
[nodemon] starting `ts-node src/worker.ts`
2024-12-09T18:27:38.879Z [INFO] Creating worker {
  sdkComponent: 'worker',
  taskQueue: 'ip-address-ts',
  options: {
   ...
  }
}
...

2024-12-09T18:27:39.412Z [INFO] Workflow bundle created { sdkComponent: 'worker', taskQueue: 'ip-address-ts', size: '0.82MB' }
2024-12-09T18:27:39.493Z [INFO] Worker state changed {
  sdkComponent: 'worker',
  taskQueue: 'ip-address-ts',
  state: 'RUNNING'
}
```

In the output, you see the Worker options and their values, the Webpack bundling output, and the state of the Worker.

> Before running this Workflow, write tests to prove it works as expected.

Author: ```suggestion
Your Worker is running and is polling the Temporal Service for Workflows to run, but before you start your Workflow, you'll write tests to prove it works as expected.
```

Reviewer: You mean "polling the Task Queue" to be a bit more specific?

```typescript
import { Connection, Client } from '@temporalio/client';
import { getAddressFromIP } from './workflows';
import { nanoid } from 'nanoid';
import { TASK_QUEUE_NAME } from './shared';
import process from 'process';

async function run() {

  if (process.argv.length <= 2) {
    console.error('Must specify a name as the command-line argument');
    process.exit(1);
  }

  const name = process.argv[2];
  const connection = await Connection.connect({ address: 'localhost:7233' });
  const client = new Client({ connection });

  const handle = await client.workflow.start(getAddressFromIP, {
    taskQueue: TASK_QUEUE_NAME,
    args: [name],
    workflowId: 'getAddressFromIP-' + nanoid(),
  });
  console.log(await handle.result());
}

run().catch((err) => {
  console.error(err);
  process.exit(1);
});

```

You import the `process` module so you can access command-line arguments you'll pass.

In the `run` function you check to see if there is at least two arguments passed. You need to check for two arguments because `process.argv` includes the Node.js executable path and the script path as the first two elements, so the actual command-line arguments start at index 2.

> The `run` function then sets up a connection to your Temporal Server, invokes your Workflow, passes in an argument for the `name`, and assigns the Workflow a unique identifier with Nanoid. The client dispatches the Workflow on the same `hello-world` Task Queue that the Worker is polling on.

Author: ```suggestion
The `run` function then sets up a connection to your Temporal Server, invokes your Workflow, passes in an argument for the `name`, and assigns the Workflow a unique identifier with Nanoid. The client dispatches the Workflow on the same Task Queue that the Worker is polling on. That's why you used a constant to ensure the Task Queue name is consistent. If there's a mismatch, your Workflow will execute on a different Task Queue and there won't be any Workers polling for tasks.
```

:::note Specify a Workflow Id

A Workflow ID is unique in a Namespace and identifies a Workflow Execution. Using an identifier that reflects some business process or entity is a good practice. For example, you might use a customer identifier as part of the Workflow Id if you run one Workflow per customer. This would make it easier to find all of the Workflow Executions related to that customer later.

In this tutorial you're generating a UUID and appending it to a string that identifies the Workflow.
:::

With the script created, open a new terminal and switch to the project directory:

```command

> cd ip-geo-ts

Author: ```suggestion
cd temporal-ip-geolocation
```

```command
npm run workflow Brian
```

You'll see the following output:

```output
> temporal-hello-world@0.1.0 workflow
> ts-node src/client.ts Brian

Hello, Brian. Your IP is 198.51.100.23 and your location is Seattle,  Washington, United States
```

Your Temporal application works. Now review it in the Temporal Web UI.

## Exploring your application in the Web UI

The Temporal Web UI gives you insights into your Workflow's execution. Open the Temporal Web UI by visiting `http://localhost:8233` and click on your completed Workflow to view the execution history. You'll see results similar to the following image:

![The results of the Workflow](images/overview.png)

You'll see the dates the Workflow Exeuction ran, how long it took to run, the input to the Workflow, and the result.

Below that, you see the Event History, detailing the entire flow, including the inputs and outputs of the Activity Executions:

![The Workflow History](images/history.png)

The most recent event is at the top, so read the history from the bottom up to see each step in the process. Using this history, you can see exactly how your Workflow executed and pinpoint any places things might have gone wrong.

> Temporal stores the results of each Activity in this history, as you can see im the image. If there was a system crash between the `getIP` and `getLocationInfo` Activity Executions, a new Worker would re-run the Workflow, but would use the previous Event History to reconstruct the Workflow's state. Instead of re-running the `getIP` function, it would use the previous run's value and continue on.

Author: ```suggestion
Temporal stores the results of each Activity in this history, as you can see im the image. If there was a system crash between the `getIP` and `getLocationInfo` Activity Executions, a new Worker would re-run the Workflow, but would use the previous Event History to reconstruct the Workflow's state. Instead of re-running the `getIP` function, it would use the previous run's value and continue on. This prevents duplicate executions. By relying on the stored Event History, Temporal ensures that the Workflow can recover seamlessly, maintaining reliability and consistency even after a crash.
```
